### DISCUSSION TITLE
Langchain4j + Arc

### DISCUSSION THREAD

**Discussion by patwlan** [langchain4j/langchain4j#2216] [Ideas]
Hi,

I am Pat, one of the lead maintiners on the open source Arc project, https://lmos-ai.github.io/arc/. 
Arc is an open source kotlin based LLM Agent Framework that is powering one of Germanys biggest LLM Chatbots.

We believe LangChain4j is a great framework and that Arc can provide an excellenet Agent abstract layer on top.

We would appreciate any feedback you might have or any ideas on how Arc can maybe enrich your amazing eco-system.

Thanks for your time.

Pat

**Comment by patwlan** [comment]
Yes, that is the big picture of what it is we are building. 

Concerning Langchain4j and Arc, we definitely see Arc at a higher level of abstraction building on top of the Langchain4j ecosystem. 
The main focus of Arc is to provide a system that can compete with the simplicity of low-code solutions and combine that with the power of a JVM stack.

Basically, what we saw over the past year is that a lot of our developers were prototyping and developing the Agents outside the JVM application, we had built, using python and low-code tools.
The problem was that they then had to integrate their solution into the actual JVM application, which was both time-consuming and error-prone.

Now with Arc, creating an Agent is now done is a few lines of code and this has really turned things around:

```
agent {
  name = “my-agent”
  prompt {
     “””
       You are a helpful assistant. You are talking to ${userProfile(“name”)}.
     ”””
  }
}
```

Besdes that, we see our focus more in providing tooling and insights into the workings of Agents, as well as, defining APIs for inter-agent communications nd the big picture that Robert expained.  
The Arc View being an example of such tooling, https://lmos-ai.github.io/arc/docs/view

And all this can be integrated into any Java based Application. There is no need to write anything outside the Agent DSL in kotlin.

I am hope that makes sense and we have not overloaded you with infos 🙂

**Reply by maxandersen** [reply]
interesting stuff - you say "any java based application" - do you mean that? I saw lots of Spring in the github repo but not sure if just part of example/templates. Is arc mainly standalone on top of langchain4j?

p.s. really wish it wasn't same name as arc the dependency container used in Quarkus but hard to name things ;)

**Reply by RobWin** [reply]
@maxandersen Yes, we are using Spring Boot right now at DT.
But Arc itself can also be used in (integrated into) Quarkus or Micronaut. The core is framework-agnostic.

**Reply by RobWin** [reply]
> BTW Arc agents seems to be a very similar concept to [AI Service](https://docs.langchain4j.dev/tutorials/ai-services) in LC4j

Yes and no, because in Arc you can use the full power of Kotlin in System Messages. Annotations are not flexible enough in real-live scenarios.  If you like we can also have a meeting someday and exchange our experiences. 

We even started with a Templating Engine for Prompts, but they weren't flexible enough as well.  

But other than that, yes, AI Services would be the best comparison.

**Reply by RobWin** [reply]
For quick prototyping you can also run the Kotlin DSL in a Kotlin Script with hot-reload and no compilation. This also helped a lot.

**Reply by kpavlov** [reply]
For prototyping one may also use Kotlin Notebooks.
As for template engine - the [PR](https://github.com/langchain4j/langchain4j/pull/2194) in JC4j is sitting and waiting, along with [other](https://github.com/kpavlov/langchain4j-kotlin/blob/main/docs/PromptTemplates.md) POCs.

**Reply by RobWin** [reply]
Our prototypes are full fletched Spring Boot applications with a UI, metrics and logs.
Just run a JBang command  `arc run agent_script.kts` and a Spring Boot app is downloaded and the Agent is started within. Including ChatUI, Metrics and more. It's not just a Notebook script we are talking about. It should be a close-to-production prototype.

**Reply by maxandersen** [reply]
just spotted the use of jbang (nice!) but hit this issue https://github.com/lmos-ai/arc/issues/53

**Reply by patwlan** [reply]
Thanks, really appreciate it! 

You can also run:
- arc new my-agent
- arc run agents

that will create a new folder "agents" containing a default agent that you can modify.

**Comment by kpavlov** [comment]
[Here](https://github.com/kpavlov/langchain4j-kotlin/blob/main/samples/src/main/kotlin/me/kpavlov/langchain4j/kotlin/samples/OpenAiChatModelExample.kt) is the PoC for ChatLanguageModel Kotlin API:
```kotlin
val response = model.chatAsync {
        messages += systemMessage("You are a helpful assistant")
        messages += userMessage("Say Hello")
        parameters {
            temperature = 0.1
            modelName = "gpt-4o-mini"
        }
    }
println("AI Answer: \"${response.aiMessage().text()}\"")
```

**Reply by dliubarskyi** [reply]
Nice!

**Comment by patwlan** [comment]
Soooo, I couldn't help noticing that “LangChain4j does not support high-level abstractions like "agent" in AutoGen or  CrewAI to build multi-agent systems.” (https://docs.langchain4j.dev/tutorials/agents#multi-agent)

I know this may be a bold question, but what would it take to get a mention of Arc in that paragraph? 
Especially since we build a lot on top of LangChain4j  🙂

If it helps sweeten the deal, we will be at KotlinConf 2025 and JavaLand and happy to give LangChain4j a shout out 🙂
(which we would probably do anyway)
I was thinking about making a joke about the name, @maxandersen would it be ok to use your quote “...wish it wasn't same name as arc…“”?

**Reply by RobWin** [reply]
Eclipse LMOS Arc :)
[Eclipse LMOS](https://eclipse.dev/lmos/) is the platform to build multi-agent systems. Arc is the Kotlin framework to build Agents, which then can be composed into a larger system.

**Reply by dliubarskyi** [reply]
Hey, is there an example of building a multi-agent system with Arc?

**Reply by RobWin** [reply]
Hello,
if you like we can also meet in Discord and I can give you an introduction into Eclipse LMOS and Arc.
Will you be on JavaLand this year? I will give a talk about LMOS there.
Here are some pictures, but there is so much more to tell about Agent/Tool Description Format, Agent/Tool Discovery, Agent Lifecycle Mangement, Canary-Release Deployments, Agent Grouping, Multi-Tenancy, Multi-Channel Support, Routing, ...
Inside of Arc we can use Langchain4j to access LLMs as an alternative to our own Kotlin-based implementation

![image (19)](https://github.com/user-attachments/assets/ba3a32aa-d2d6-4ea9-b371-1261ec51cce8)
![image (20)](https://github.com/user-attachments/assets/6069d8c7-3cb6-4290-bec4-2ffd7cb3a523)
![image (18)](https://github.com/user-attachments/assets/435b8167-8355-419e-86a0-d32f547b6c08)
![image (21)](https://github.com/user-attachments/assets/3a6a97d3-4c74-4519-b6c1-d1dc49ae71ab)

**Reply by LizeRaes** [reply]
@RobWin if this meeting didn't take place yet, I would be interested to listen in too.

**Reply by kpavlov** [reply]
@RobWin , I would love to join too

**Reply by RobWin** [reply]
Next Friday at 9 AM CET, I'll be presenting Eclipse LMOS, LMOS Protocol and Web of Things in a joint call with three W3C Community Groups: WebAgents, Web of Things, and the WebThing Protocol Community Group.

You can find the event details here: https://www.w3.org/events/meetings/5a77b997-467b-4d5d-bb40-b84789c17d2e/20250314T090000/

I will explain how we use Eclipse LMOS at Deutsche Telekom and why WoT can be applied for modeling Agents and Tools. I also hope to address questions. Plus, several other W3C WoT experts will be available for insights and discussion.

**Reply by RobWin** [reply]
If you like you can also join our Discord. We can also have discussions there. We would really like to know your opinion on how we could more deeply integrate Langchain4j into LMOS and Arc.
Or how we could apply your Kotlin Extensions @kpavlov in our Langchain4j Integration.

https://discord.gg/yxG8h9pZ

**Reply by kpavlov** [reply]
@RobWin, I have backported streaming ChatLanguageModel api from langchain4j-kotlin to langchain4j. Suspend functions api for document loaders is [pending](https://github.com/langchain4j/langchain4j/pull/2640). Please, let me know which Kotlin api for lc4j you are missing the most, so I can prioritize migration. Async api might become available via langchain4j-kotlin library very quickly.

**Reply by RobWin** [reply]
Recording is here 

https://youtu.be/gcpHYtqTBbA?si=UGe9f2uUWJO4tkGP

**Comment by smileyboy2019** [comment]
Similar to Microsoft's AutoGen, it is planned to be implemented in that version.
His technical principles
1. Define a team and set multiple members
2. Implement the start and end actions of the task
3. Sequential execution of evaluation of the previous large model
4. When the loop condition or trigger condition is reached, it ends.

**Comment by patwlan** [comment]
Hi, we just updated the landing page for Arc https://eclipse.dev/lmos/arc2/index.html. Any feeback would be appreciated.
... so i dont want to be a nuisance, but any decision on adding a reference to Arc here https://docs.langchain4j.dev/tutorials/agents#multi-agent? no hard feelings either way... :)
